### PR TITLE
Align to BIP78

### LINKED ISSUE #2: Adjust to be inline with BIP78
Will try and go through it today and see what exactly needs adjustment

### PR DESCRIPTION
**Pull request by Kukks**
closes #2

This removes the address format restriction, and adds the new optional parameters and checks associated with that.

### CODE REVIEW

**Review by junderw — CHANGES_REQUESTED** [review]
Please fix the tests. CI should be passing, and ideally we should be upping the coverage threshold to 90 across the board.

Basically, by reading the test titles "describe x, it should y" we should be able to get an idea that all the points in the BIP are covered.

From a cursory glance, everything looks ok. But I can't merge this without tests passing at least.

**Comment by Overtorment** [comment]
@Kukks tests still fail

**Comment by junderw** [comment]
After fixing the format problem and the build problem it seems as though tests are still failing.

**Comment by Overtorment** [comment]
attempt to use this branch in bluewallet throws  `Can not add duplicate data to input`

```
overtorment@laptop3 ~/Documents/BlueWallet (master) $ ./node_modules/.bin/jest     tests/unit/payjoin-transaction.test.js
 FAIL  tests/unit/payjoin-transaction.test.js
  PayjoinTransaction
    ✓ throws if smth is wrong with pj transaction (47ms)
    ✕ works (6ms)

  ● PayjoinTransaction › works

    Can not add duplicate data to input

      at Object.updateInput (node_modules/bitcoinjs-lib/node_modules/bip174/src/lib/utils.js:99:19)
      at Psbt.updateInput (node_modules/bitcoinjs-lib/node_modules/bip174/src/lib/psbt.js:46:13)
      at Psbt.updateInput (node_modules/bitcoinjs-lib/src/psbt.js:577:15)
      at PayjoinClient.run (node_modules/payjoin-client/src/client.js:165:23)
```

https://github.com/BlueWallet/BlueWallet/pull/1926

**Comment by junderw** [comment]
I have pushed a new image to `junderw/btcpay-client-test-server`

Uses latest NBXplorer and BTCPayServer tagged versions.

**Comment by junderw** [comment]
Integration tests are failing because it's trying to fetch('bitcoin:xxx') and fetch needs an http/https link.

So the issue is in the parsing of the URL.

The URL parsing logic is very weird.

If the URL has one logic and the optional opts has another logic it looks as though it will overwrite the URL logic with the opts logic.

Since the URL represents what the server wants to do, if what WE want to do (opts) is incompatible it should throw an error.

**Comment by Kukks** [comment]
> Integration tests are failing because it's trying to fetch('bitcoin:xxx') and fetch needs an http/https link.
> 
> So the issue is in the parsing of the URL.
> 
> The URL parsing logic is very weird.
> 
> If the URL has one logic and the optional opts has another logic it looks as though it will overwrite the URL logic with the opts logic.
> 
> Since the URL represents what the server wants to do, if what WE want to do (opts) is incompatible it should throw an error.

Yeah, the situation can be a little confusing:
The BIP21 payload can signal various things for BIP78 payjoin, which would set some of the optional parameters when communicating with the payjoin server. IMO, the payjoin client should be able to override these if they really wanted though.



> I have pushed a new image to `junderw/btcpay-client-test-server`
> 
> Uses latest NBXplorer and BTCPayServer tagged versions.

Thanks for the bump, will start fixing this part next. I fixed the rest of the spec tests and added a few more, but I may need to add more since coverage may be a bit down now.

**Comment by junderw** [comment]
>  the payjoin client should be able to override these if they really wanted though.

If this is allowed it should be in the BIP. These kind of implementation details of "oh, the server requires you follow X but the client implementation allows people to override X which will cause a server error" are exactly what BIPs are designed to avoid.

**Comment by Kukks** [comment]
> > the payjoin client should be able to override these if they really wanted though.
> 
> If this is allowed it should be in the BIP. These kind of implementation details of "oh, the server requires you follow X but the client implementation allows people to override X which will cause a server error" are exactly what BIPs are designed to avoid.

This was a misunderstanding. In my refactoring, I thought I had started requiring the full BIP21 instead of just the payjoin endpoint, but I was wrong. I've changed it now and the tests seem to be passing except for the 65% coverage quota (currently 60)

**Comment by junderw** [comment]
There's quite a few branches missing.

I added html reporter to this branch, so run `npm run coverage` locally then open ./coverage/index.html in a web browser to get an idea of which branches are not covered...

It looks like a ton of the branches are major logic paths, and critical to test before release.

Ideally coverage should be >=90% across the board, but luckily it looks like most of the missing lines and statements are inside these branches, so if we get branches above 90% I think everything else will be over 90 too.

**Comment by junderw** [comment]
Published 9241a826768b673f8d371f2d60db975d4139cccf as tag `beta` on npm. The version is `1.0.0-beta`

File: ts_src/client.ts [hunk 1]
```diff
@@ -1,55 +1,58 @@
...
+import { PsbtInput, PsbtOutput } from 'bip174/src/lib/interfaces';
+
+const BROADCAST_ATTEMPT_TIME = 2 * 60 * 1000; // 2 minute
 
-const BROADCAST_ATTEMPT_TIME = 1 * 60 * 1000; // 1 minute
+export interface PayjoinClientOptionalParameters {
+  disableOutputSubstitution?: boolean;
+  payjoinVersion?: number;
+  additionalfeeoutputindex?: number;
+  maxadditionalfeecontribution?: number;
```
Comment (junderw) [inline ts_src/client.ts]: camelCase

File: ts_src/utils.ts [hunk 2]
```diff
@@ -200,3 +93,58 @@ export function getInputIndex(
...
+}
+
+export function getEndpointUrl(
+  url: string,
+  payjoinParameters?: PayjoinClientOptionalParameters,
+): string {
+  if (!payjoinParameters) {
+    return url;
+  }
+  const parsedURL = new URL(url);
```
Comment (junderw) [inline ts_src/utils.ts]: Don't use URL.

**Review by junderw — CHANGES_REQUESTED** [review]

**Comment by junderw** [comment]
Published 98f0e3b as tag `beta` on npm. The version is `1.0.0-beta2`

**Comment by Overtorment** [comment]
querystring is also apparently very limited in RN. can `encodeURIComponent` be used instead?

**Comment by Overtorment** [comment]
just tested latest commit with bw, worked okay!

**Comment by junderw** [comment]
Because we had to roll our own method for parsing querystrings, we need WAAAAAAAAAAY more tests. Test a bunch of edge cases. What if the server is malicious and tries to somehow break our wallet by putting in a bad URL with tons of ? or weird escaping loops (an escape of an escape of an escape of an escape symbol etc.)... String parsing is one of the most overlooked sources for vulnerabilities, and the URL is supplied by someone we are assuming is evil in other parts of the library (we check everything they tell us on our side, but URL is parsed blindly) We probably should throw some errors in certain cases of malformed URLs.

**Comment by Kukks** [comment]
Ok ok, what if we revert to the original, and let BW provide a qs manager
override

**Comment by junderw** [comment]
That would be fine.

**Comment by Overtorment** [comment]
I have just tested BW with lates payjoin-client from this PR, and everything worked well
image

**Review by junderw — APPROVED** [review]
LGTM.

We will publish under a major version bump.

**Comment by junderw** [comment]
published to npm as v1.0.0

**Comment by Overtorment** [comment]
so far so good

https://github.com/BlueWallet/BlueWallet/pull/2203


image

image

**Comment by junderw** [comment]
please also try some of the newer payjoin features if you can.

**Comment by Overtorment** [comment]
how do I do that? I know only one payjoin-enabled server https://btc.donate.kukks.org
